lib/managers/files.js

```diff
@@ -1071,81 +1067,51 @@ Files.prototype.commitUploadSession = function(sessionID, fileHash, options, cal
 			}
 		};
 
-	var self = this;
-	async.waterfall([
-		function getParts(waterfallCallback) {
-
-			if (parts && parts.length) {
-				waterfallCallback(null);
-				return;
-			}
-
-			var nextMarker;
-
-			async.doWhilst(
-				function fetchParts(untilCallback) {
+	var fetchParts = (offset, fetchedParts) => {
 
-					let pagingOptions = {
-						limit: 1000
-					};
-
-					if (nextMarker) {
-						pagingOptions.marker = nextMarker;
-					}
+		let pagingOptions = {
+			limit: 1000,
+			offset: offset
+		};
 
-					self.getUploadSessionParts(sessionID, pagingOptions, function(err, data) {
+		return this.getUploadSessionParts(sessionID, pagingOptions)
+			.then(data => {
 
-						if (err) {
-							untilCallback(err);
-							return;
-						}
+				fetchedParts = fetchedParts.concat(data.entries);
 
-						nextMarker = data.next_marker;
-						parts = parts.concat(data.parts);
-						untilCallback(null);
-					});
-				},
-				() => typeof nextMarker === 'string' && nextMarker.length > 0,
-				function finalizeFetchParts(err) {
+				if (data.offset + data.entries.length >= data.total_count) {
+					return Promise.resolve(fetchedParts);
+				}
 
-					if (err) {
-						waterfallCallback(err);
-						return;
-					}
+				return fetchParts(offset + data.limit, fetchedParts);
+			});
+	};
 
-					waterfallCallback(null);
-				});
-		},
-		function commitUpload(waterfallCallback) {
+	return (userParts ? Promise.resolve(userParts) : fetchParts(0, []))
+		.then(parts => {
 
 			params.body.parts = parts;
-			self.client.post(apiURL, params, waterfallCallback);
-		}
-	], function(err, response) {
+			return this.client.post(apiURL, params);
```

Comment:
It was kind of nice to have the named "commitUpload" function previously. Could you just add a comment here? e.g. // Commit upload